ILL loaders refactoring

darominski opened the pull request:
**Description of work.**

This PR implements solutions to found issues found during the investigation summarised in the related issue (#34499):

- All ILL loaders now interface to two functions loading data into a workspace, one used for a static workspace and the other for a moving workspace, both hosted in `LoadHelper` namespace
- All loaders now consistently use `int` data, instead of a mix of `uint` and `int`
- Scan data type for IN5, PANTHER, and SHARP is now loaded by `LoadILLTOF2` along with the rest of data types for these instruments, rather than being loaded by `LoadILLDiffraction`
- Instrument is always loaded via single function hosted in `LoadHelper`
- Improved unit test coverage
- IN4 monitor now has a more reasonable ID, in line with other ILL TOF instruments

**To test:**
1. Code review.
2. Consistent behaviour is ensured by existing and added unit and system tests.

Fixes #34499 . 

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
[ILL] Generalized loading empty instrument to a LoadHelper function

This adds a new method for loading an empty instrument to a provided workspace
in LoadHelper class. All ILL loaders now refer to that method instead of creating
individual children algorithms.

darominski committed:
Changed LoadHelper from class to namespace

Previously, LoadHelper was a class with no specialised constructor and no meaningful objects, with all classes
being effectively static. Therefore, it was already a namespace in all but name.

darominski committed:
Merge remote-tracking branch 'origin/main' into 34499_ill_loaders_refactoring

darominski committed:
Added missing DllConfig imports

which became necessary after move from DLLExport to MANTID_DATAHANDLING_DLL
flag for the classes in Mantid DataHandling.

darominski committed:
Added missing doxygen for LoadEmptyInstrument method

darominski committed:
[ILL] Monitor ID for IN4 now 100000

was 0, which would cause conflict when generalizing filling workspaces.

darominski committed:
[SANSILL] Added more asserts for loaded data

to check the exact value loaded, and also whether monitors are not placed
at the beginning of the workspace.

darominski committed:
[ILL] Generalized filling workspaces into single LoadHelper function

the fillStaticWorkspace can handle all cases of non-scan data moving from NXInt data
to MatrixWorkspace.

darominski committed:
Extracted replacing zero counts errors to its own function

in LoadHelper namespace.

darominski committed:
Merge remote-tracking branch 'origin/main' into 34499_ill_loaders_refactoring

darominski committed:
[SANSILL] Fix loading legacy D16 omega monitors

darominski committed:
[DiffILL] Added more asserts for loaded data

to check the exact value loaded, and also whether monitors are not placed
at the beginning of the workspace, and workspace type (distribution, histogram).

darominski committed:
[DirectILL] Re-organised loading monitor information

Now there is only one function that fetches paths to monitors and sets the number of monitors
based on that information. Loading monitor data is done the same way as regular data via
external LoadHelper method.

darominski committed:
LoadHelper: new method to load scan into moving workspace

darominski committed:
LoadHelper: fillStaticWorkspace uses const ref for data

darominski committed:
[DiffILL] Generalized loading into LoadHelper methods

with one method (fillStaticWorkspace) used for static scans, and another
(fillMovingWorkspace) for moving scans.

darominski committed:
[SANSILL] Simplify condition for cppcheck

for deciding whether the loaded data is a legacy D16 single-point scan
or new D16B data.

darominski committed:
[DirectILL] Removed redundant doxygen doc

from the initalisation of workspace, as monitors are now loaded in a different place.

darominski committed:
[DirectILL] Fixed IN4 tests after monitorID change

Which were breaking since the monitor ID was set to 0 and was clashing with
actual detector IDs.

darominski committed:
Simplified setting custom detector ID

darominski committed:
[DiffILL] Corrected checking for monitor in scan ws

darominski committed:
Option for custom detector IDs in moving workspace filling

darominski committed:
[DiffILL] Suppression of redundant D2B const parameter

darominski committed:
[DiffILL] Corrected comment about dimension order for loading

darominski committed:
[DiffILL] Corrected filling of D2B 'Raw' data type

as in that case the tubes are filled in a zig-zag way with even tubes
having first pixel on the top while odd tubes at the bottom of the tube.
This is not the case for calibrated data.

darominski committed:
[DiffILL] Expanded asserts on exact loaded value

darominski committed:
[DiffILL] Move to const-length vector of customDetectorIDs

darominski committed:
[DirectILL] Removed redundant calculateError function

darominski committed:
Added helper functions for getting NXInt and NXDouble data

darominski committed:
[ILL] Get NXS datasets via helper functions

darominski committed:
[DirectILL] Scan data types can now be loaded with LoadILLTOF

while previously loading of diffraction and TOF data types were loaded with LoadILLTOF,
and scans were handled by LoadILLDiffraction.

darominski committed:
[DirectILL] Doxygen fixes

added missing parameter descriptions and a default value for the scannedVarId.

darominski committed:
[DiffILL] Removed redundant start_time variable

Instead, the m_startTime object parameter is used for higher consistency.

darominski committed:
Simplified replacing zero error counts

darominski committed:
[SANSILL] Removed redundant getComponentPosition function

the calls shall be handled by the same function but hosted in LoadHelper.

darominski committed:
Remove dumpNexusAttributes method

as it is not called from anywhere in the codebase nor exposed to Python to be used in scripting.

darominski committed:
Corrected checking for empty instrumentPath

darominski committed:
Generalized filling static workspaces with arbitrary data order

based on the default of tube - pixel - channel, and searching of
the matching index in the provided axisOrder parameter.

darominski committed:
[SANSILL] Corrected loading logic, improved comments

darominski committed:
[SANSILL] Corrected behaviour for single-scan D16B data

darominski committed:
[SANSILL] Unit test for single-point D16B scan

darominski committed:
Merge remote-tracking branch 'origin/D16B_single_scan_fix' into 34499_ill_loaders_refactoring

darominski committed:
Merge remote-tracking branch 'origin/main' into 34499_ill_loaders_refactoring

darominski committed:
[REFLILL] Removed double-loading of monitor data

darominski committed:
Merge remote-tracking branch 'origin/main' into 34499_ill_loaders_refactoring

darominski committed:
Allowed for parallelization of filling workspaces

perenon commented on docs/source/algorithms/LoadILLDiffraction-v1.rst:
```diff
@@ -119,6 +119,8 @@ Output:
 
   The output has 1 bin (detector scan) and 1281 spectra
 
+<<<<<<< HEAD
```
I think these are trace of previous merge conflict. Can they be fixed ?

darominski replied:
Absolutely, will fix it immediately. At the same time, there's an obsolete IN5 example that somehow got stuck there, so I am going to remove them both simultaneously.

perenon commented on Framework/DataHandling/src/LoadHelper.cpp:
```diff
@@ -353,62 +355,6 @@ void LoadHelper::recurseAndAddNexusFieldsToWsRun(NXhandle nxfileID, API::Run &ru
 
 } // recurseAndAddNexusFieldsToWsRun
 
-/**
```
I guess you remove dumpNexusAttributes because it is unused in the code ?

darominski replied:
Indeed, it is not used anywhere in the codebase for quite a long time. There were parts of the code that depended on HDF4.

perenon commented on Framework/Algorithms/test/CorrectTOFAxisTest.h:
```diff
@@ -231,7 +233,7 @@ class CorrectTOFAxisTest : public CxxTest::TestSuite {
     auto alg = createCorrectTOFAxisAlgorithm();
     TS_ASSERT_THROWS_NOTHING(alg->setProperty("InputWorkspace", inputWs))
     TS_ASSERT_THROWS_NOTHING(alg->setPropertyValue("OutputWorkspace", "_unused_for_child"))
-    TS_ASSERT_THROWS_NOTHING(alg->setPropertyValue("IndexType", "Workspace Index"))
```
Does that mean the test was not properly set and you fixed it ?

darominski replied:
After the change of IN4 indexing, with monitor NOT having index 0 but a high number consistent with other ILL TOF instruments, a change had to be be made to stay consistent, and the spectrum number was a better choice.

perenon replied:
Sorry if that is a bit pedantic, but I try to figure out what could affect other algorithms.
 
If I get it right, the purpose of these tests seems to be the control that the algorithm is correctly executed if parameters are OK, and not being executed if parameters are flawed, and you also ensure you can set correctly properties, even if the global set of properties gives are incorrect (Ei is missing, here).

Since we do not test the "Workspace Index" case anymore, we must ensure this will not be used.
=> The underlying algorithm "CorrectTofAxis" is only using by DirectILLCollectData, so that is why everything is safe?

darominski replied:
That's perfectly fine. The workspace index can still be tested, but now there is an offset of 1 between the two options for IN4 due to the change mentioned before. I can easily bring back setting IndexType to Workspace Index for some tests so that both are covered. I am going to push it in the next commit.

perenon replied:
If that is possible, yes please. I am always afraid suppressing test cases. 
For IN4, I think what you did increase clarity of code !

darominski replied:
That's pushed now, in those modified cases the `ReferenceSpectra` changed as mentioned.

darominski committed:
Test covers again setting IndexType to Workspace Index

which was unset by previous changes and moved to only using Spectrum Number.

darominski committed:
[DiffILL] Removed remnants of merging from docs

and example of loading IN5 data via LoadILLDiffraction which is no longer the case.

perenon reviewed (approved):
The work proposed here is a major clarification of loaders for ILL instruments.
 
It provides corrections of the documentation, increasing in test coverage and most importantly code factorization for ILL loaders.

Considering non-ILL specific code, method `dumpNexusAttributes` is removed (it is unused) from `LoadHelper.` Two new methods are notably added `fillStaticWorkspace` and `fillMovingWorkspace`. `CorrectTOFAxisTest` is slightly modified to take into account the fact that monitor index is no more pinned to 0 for ILL IN4 instrument.
